pinot-server/src/main/java/org/apache/pinot/server/starter/helix/BaseServerStarter.java

```diff
@@ -363,6 +377,41 @@ private void startupServiceStatusCheck(long endTimeMs) {
         ServiceStatus.getStatusDescription());
   }
 
+  /**
+   * Recursively deletes all data directories starting with tmp- last modified before the startTime.
+   * @param startTime start time of the application
+   * @param dataDir data directory to start from
+   */
+  @VisibleForTesting
+  public static void deleteTempFilesSinceCutoffTime(long startTime, @Nonnull File dataDir) {
+    if (!dataDir.exists() || !dataDir.isDirectory()) {
+      LOGGER.warn("Data directory {} does not exist or is not a directory", dataDir);
+      return;
+    }
+    IOFileFilter beforeStartTimeFilter = new AgeFileFilter(startTime, true);
+    IOFileFilter tmpPrefixFilter = new PrefixFileFilter("tmp-");
```

Comment:
The prefix could be an argument to this method?

Reply:
it could, but that felt like a better and easy addition for a future PR that requires it

Reply:
would this delete data from table whose name has `tmp-` prefix? because table segments are put under dir like `<dataDir>/<tableNameWithType>/...` as initialized [here](https://github.com/apache/pinot/blob/master/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/data/manager/TableDataManagerConfig.java#L79): 
```
defaultConfig.addProperty(TABLE_DATA_MANAGER_DATA_DIRECTORY,
        instanceDataManagerConfig.getInstanceDataDir() + "/" + tableNameWithType);
```

Reply:
Right, these are the side effects that would be good to minimize. Perhaps it can be a config too?

Reply:
hmm.. I gave it a try in my test env and found I could create tables with 'tmp-' prefix. Things like pushing segment to it and querying it worked as usual. In the server's dataDir, there was folder for the table like below. So it may be risky to use "tmp-" to clean up dirs. 

```
...PinotServerDataDir0/tmp-baseballStats_OFFLINE/tmp-baseballStats_OFFLINE_0/v3/
columns.psf          creation.meta        index_map            metadata.properties
```

![image](https://user-images.githubusercontent.com/61892277/148141545-5000a8dd-abd2-4cde-9bc0-4379b7469adc.png)

Reply:
Thanks for pointing out! If that's the case, can we revisit the logic of creating the `tmp-` directory inside the data directory? Can we move it anywhere else?

Reply:
> We may extract tmp dir creation and cleanup logic to a util class to manage its naming convention and location in one place, in case later on, folks use tmp. or temp- prefix or put them outside dataDir, making your cleanup logic less effective.

This is all fair feedback. I thought about this and figured it was unlikely people were creating `tmp-` tables, but let's not leave it to chance. It seems we have 2 options:
1) encode the tmp segment name with something that cannot be used for real table names. The only thing disallowed in code is double underscore: `// Double underscore is reserved for real-time segment name delimiter`
2) put tmp segments in a different directory

2 seems like the better option given that's how most uses of tmp things work in typical software. I believe our issue does stem from `BaseTableDataManager`, so how about we add a new `getTmpSegmentDataDir` function that puts things in `File(_indexDir, "tmp", segmentName);`? At that point, can we just clear that directory every time the process comes up regardless of age?

Reply:
Thanks for the proposals. I prefer to option 2 as well.

There are a few places creating tmp- dir in pinot/core/data/manager module. Your util method may be used to unify all those cases. Right now, their dir suffix are unique, which (luckily) could help one figure out where deletion fails. So for the new util method, I'd suggest to allow caller to customize some part of the dir name like its suffix to help debugging. 

```
pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java:    File tempRootDir = getSegmentDataDir("tmp-" + segmentName + "-" + UUID.randomUUID());
pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/HLRealtimeSegmentDataManager.java:          File tempSegmentFolder = new File(_resourceTmpDir, "tmp-" + System.currentTimeMillis());
pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java:      File tempSegmentFolder = new File(_resourceTmpDir, "tmp-" + _segmentNameStr + "-" + now());
pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java:    File tempSegmentDir = new File(_indexDir, "tmp-" + segmentName + "." + System.currentTimeMillis());
```